---
title: "Transposon mutagenesis create summary files"
author: "Grace Avecilla"
date: "February 10, 2020"
output:
  html_notebook:
    toc: yes
  html_document:
    df_print: paged
    toc: yes
---
This notebook will take files generated on the HPC and create summary Rdata files to be used for plotting in Hermes_mutagenesis_code_draft.Rmd

#Set up
```{r set up, echo=F}
library(knitr)
library(kableExtra)
require(GGally)
library(ggrastr)
library(cowplot)
library(clusterProfiler)
library(org.Sc.sgd.db)
library(DESeq2)
library(tidyverse)
library(OneR)
library(EnhancedVolcano)
library(labtools)

select <- dplyr::select

dir = "/Volumes/GoogleDrive/My Drive/Gresham Lab_Grace/greshamlab_github/hermes_analysis/"
setwd(dir)

seq_runs = c("bgi1","bgi2", "nyc1", "nyc2")
```

## CNV structure (by read depth) before and after experiment




## Insertion positions

First, I have to get the number of insertion positions and the number of reads per insertion from each sequencing run, and combine them.
```{r reads per position}
# function to read in data, add sample label
get_rpp = function(x, ver) {
  filepath=paste0(dir,"hpc_output/",ver,"/",x)
  out = read_tsv(filepath, col_names = F, col_types = cols()) %>%
    dplyr::rename(chromosome=X1, chr_pos=X2, reads=X3) %>%
    mutate(sample = str_sub(x, 1, -16),
           version = ver)
}

#read in data from each sequencing run
rpp_df_each = NULL
for(v in seq_runs) {
  files = list.files(path = 
                       paste0(dir, 'hpc_output/', v,'/'),
                     pattern = '*readPerPos.txt')
  read_per_pos = map(files, get_rpp, ver=v)
  df = do.call(rbind, read_per_pos)
  rpp_df_each = rpp_df_each %>% bind_rows(df)
}

#read in the data from all runs combined
files = list.files(path = paste0(dir, 'hpc_output/combined/'),
                     pattern = '*readPerPos.txt')
read_per_pos = map(files, get_rpp, ver='combined')
rpp_df = do.call(rbind, read_per_pos) %>%
  mutate(chromosome = case_when(chromosome == 'NC_001133.9' ~ "1",
                                             chromosome == 'NC_001134.8' ~ "2",
                                             chromosome == 'NC_001135.5' ~ "3",
                                             chromosome == 'NC_001136.10' ~ "4",
                                             chromosome == 'NC_001137.3' ~ "5",
                                             chromosome == 'NC_001138.5' ~ "6",
                                             chromosome == 'NC_001139.9' ~ "7",
                                             chromosome == 'NC_001140.6' ~ "8",
                                             chromosome == 'NC_001141.2' ~ "9",
                                             chromosome == 'NC_001142.9' ~ "10",
                                             chromosome == 'NC_001143.9' ~ "11",
                                             chromosome == 'NC_001144.5' ~ "12",
                                             chromosome == 'NC_001145.3' ~ "13",
                                             chromosome == 'NC_001146.8' ~ "14",
                                             chromosome == 'NC_001147.6' ~ "15",
                                             chromosome == 'NC_001148.4' ~ "16",
                                chromosome == 'NC_001224.1' ~ "mito"))
```

### Summary of reads per library and reads per position - Supplementary Table???
Now get summary data for reads per position for each library.
```{r summary reads per position for each individual library}
rpp_df_each %>% group_by_at(vars(version, sample)) %>% 
  summarize(total_sites = n(), min_rpp = min(reads), max_rpp = max(reads),
            mean_rpp = mean(reads), median_rpp = median(reads)) %>%
  kable() %>%
  kable_styling()
```
Now get summary data for reads per position when all BAMs are combined.
```{r summary reads per position for all combined}
summary_rpp = rpp_df %>% group_by_at(vars(sample)) %>% 
  summarize(total_sites = n(), min_rpp = min(reads), max_rpp = max(reads),
            mean_rpp = mean(reads), median_rpp = median(reads))
summary_rpp %>%
  kable() %>%
  kable_styling()

```

## Insertions per gene
Note, insertions per gene here is the number of unique insertion sites per gene, not taking into account the number of reads per each insertion site.

First, I have to get the number of insertion positions and the number of reads per insertion from each sequencing run, and combine them.
```{r insertions per gene}
get_ipg = function(x, ver){
  data = read_tsv(paste0(dir, "/hpc_output/",ver, "/", x), 
                  col_types = cols(CDS=col_character(),`#insertion` = col_double())) %>%
    mutate(sample = str_sub(x, 1, -22), gene=NA,
           run=ver)
}

ipg_forcor = NULL
for(i in seq_runs) {
  ipg_files = list.files(path = paste0(dir, "/hpc_output/", i, "/"), pattern = '*insertionPerGene.txt')
  ipg_t = map(ipg_files, get_ipg, ver=i)
  ipg_df_t = do.call(rbind, ipg_t)
  ipg_forcor = ipg_forcor %>% 
    bind_rows(ipg_df_t %>% unite(sample_rep, sample, run))
}

#get all
ipg_files = list.files(path = paste0(dir, "hpc_output/combined/"), pattern = '*insertionPerGene.txt')
ipg_all = map(ipg_files, get_ipg, ver="combined")
ipg_df = do.call(rbind,ipg_all) %>%
  dplyr::rename(insertions_per_gene=`#insertion`)

# add in systematic gene names
ipg_df$gene = yeast_r64_to_systematic(ipg_df$CDS)
ipg_df = ipg_df %>% mutate(gene = if_else(CDS == "ID= GRESHAMGFP", "CNV_reporter", gene))
```

Normalized insertions per gene (insertions/kb/gene). This was already done on the HPC during the initial part of the pipeline. 
```{r get insert/kb/gene}
#get data
get_ipkbpg = function(x, ver = "combined") {
  data = read_tsv(paste0(dir, '/hpc_output/', ver, '/', x), 
                  col_names = F, col_types = cols()) %>%
    dplyr::rename(id=X1, inserts_kb_gene = X2) %>%
    mutate(sample = str_sub(x, 1, -27), gene=NA)
  data$gene = yeast_r64_to_systematic(data$id)
  data = data %>% mutate(gene = if_else(id == "ID= GRESHAMGFP", "CNV_reporter", gene))
  data %>% dplyr::select(-id)
}
ipkbpg_df = do.call(rbind, map(list.files(paste0(dir, '/hpc_output/combined'), 
                                       pattern = '*insertionPerKbPerGene.txt'),
                            get_ipkbpg, ver="combined"))
#reformat
ipkbpg = ipkbpg_df %>% 
  pivot_wider(names_from = sample, values_from = inserts_kb_gene) %>%
  mutate_all(~replace(., is.na(.), 0))
```



#### This next section needs to be fixed ####

Plot correlation between samples sequenced at bgi first time, bgi second time, and in nyc
```{r plot seq rep insert/gene}
#first get data into the right format for plotting
ipg_forcor = ipg_forcor %>%
  dplyr::select(-gene) %>%
  pivot_wider(names_from = sample_rep, values_from = `#insertion`)
# make scatter plots
# thank you Chris
# GGally plot wrapper for geom_point_rast
lower.plotter <- function(data, mapping, ...) {
  mapping[["shape"]] <- factor(16)
  mapping[["alpha"]] <- factor(0.1)
  mapping[["size"]] <- factor(1)
  mapping[["color"]] <- "black"
  ggplot(data, mapping) +
    ggrastr::geom_point_rast(...)
}

#### THIS IS NOT WORKING???####
gene.rank.plt <- GGally::ggpairs(ipg_forcor %>% dplyr::select(-CDS), mapping=aes(alpha=0.01), axisLabels = "none", diag='blankDiag', switch='both', 
                                 upper = "blank",
                                 lower = list(continuous = lower.plotter)) +
  theme(strip.background = element_blank(), strip.text = element_text(size=10))
# Shamelessly stolen from https://stackoverflow.com/questions/42654928/how-to-show-only-the-lower-triangle-in-ggpairs
ggpairs_nodiag <- function(g){
  g$plots <- g$plots[-(1:g$nrow)]
  g$yAxisLabels <- g$yAxisLabels[-1]
  g$nrow <- g$nrow -1
  
  g$plots <- g$plots[-(seq(g$ncol, length(g$plots), by = g$ncol))]
  g$xAxisLabels <- g$xAxisLabels[-g$ncol]
  g$ncol <- g$ncol - 1
  
  g
}
gene.rank.plt <- ggpairs_nodiag(gene.rank.plt)
####
gene.cor <- cor(ipg_forcor[,2:ncol(ipg_forcor)], method = "spearman", use = "pairwise.complete.obs")
gene.cor[base::lower.tri(gene.cor)] <- NA
gene.cor <- reshape2::melt(gene.cor, na.rm=TRUE, value.name = "Correlation")
gene.cor$Var2 <- factor(gene.cor$Var2, levels=rev(levels(gene.cor$Var2)))
SFig1.2a <- ggplot(gene.cor, aes(x=Var1, y=Var2, fill=Correlation)) +
  scale_fill_gradient2(low='blue', mid="white", high='red', midpoint=0.75, limit=c(0.5,1)) + 
  geom_tile() +
  geom_text(aes(label=round(Correlation,3))) +
  theme_classic() +
  labs(fill="Spearman\nCorrelation", title = "Insertions per gene") +
  theme(axis.title = element_blank(), axis.text.x = element_text(angle = 45, vjust=1, hjust=1))
```

### Supplementary table?
```{r summary insertions per gene}
insert_per_gene_summary = ipg_df %>% group_by(sample) %>% 
  summarise(total_gene = n(), min_insertions_per_gene = min(insertions_per_gene),
            max__insertions_per_gene = max(insertions_per_gene),
          mean_insertions_per_gene = mean(insertions_per_gene),
          median_insertions_per_gene = median(insertions_per_gene),
          total_insertions_in_genes = sum(insertions_per_gene))
insert_per_gene_summary %>%
  kable() %>%
  kable_styling()
```

Get total number unique insertions in coding vs non-coding regions.
```{r plot number unique insertions coding vs non-coding}
coding_v_non = insert_per_gene_summary %>% select(sample, total_insertions_in_genes) %>%
  full_join((summary_rpp %>% ungroup() %>% select(sample, total_sites)), 
            by='sample') %>% 
  mutate(total_insertions_noncoding = total_sites-total_insertions_in_genes) %>%
  distinct() %>%
  pivot_longer(cols=-sample, names_to="location_type")

#### write data for Figure 1C####
write_csv(coding_v_non, "./figure_data/Fig1C.csv")
```

### Look at insertion coverage over each genome to check that nowhere is depleted in insertions

#### HERE, do 500 or 1kb windows instead ####
```{r Insertion coverage over genome}
get_window_cov = function(x, ver) {
  data = read_tsv(paste0(dir, "/hpc_output/",ver, "/", x), col_names = F) %>%
    mutate(sample = str_sub(x, 1, -22))
}

#read in data from each sequencing run
files = list.files(path = 
                       paste0(dir, 'hpc_output/combined/'),
                     pattern = '*_insertionWindows.cov')
wind_cov = map(files, get_window_cov, ver="combined")
cov_df = do.call(rbind, wind_cov) %>%
  dplyr::rename(chromosome = "X1", bin_start = "X2", bin_end = "X3", coverage = "X4") %>%
  mutate(chromosome = case_when(chromosome == 'NC_001133.9' ~ "1",
                                             chromosome == 'NC_001134.8' ~ "2",
                                             chromosome == 'NC_001135.5' ~ "3",
                                             chromosome == 'NC_001136.10' ~ "4",
                                             chromosome == 'NC_001137.3' ~ "5",
                                             chromosome == 'NC_001138.5' ~ "6",
                                             chromosome == 'NC_001139.9' ~ "7",
                                             chromosome == 'NC_001140.6' ~ "8",
                                             chromosome == 'NC_001141.2' ~ "9",
                                             chromosome == 'NC_001142.9' ~ "10",
                                             chromosome == 'NC_001143.9' ~ "11",
                                             chromosome == 'NC_001144.5' ~ "12",
                                             chromosome == 'NC_001145.3' ~ "13",
                                             chromosome == 'NC_001146.8' ~ "14",
                                             chromosome == 'NC_001147.6' ~ "15",
                                             chromosome == 'NC_001148.4' ~ "16",
                                chromosome == 'NC_001224.1' ~ "mito"))

#### write data for FigS1.2B####
write_csv(coding_v_non, "./figure_data/FigS1.2B.csv")
```

### Find which genes in each library that have no insertions
This is already done on the HPC.

```{r get genes with no insertions}
all_genes = read_csv("/Volumes/GoogleDrive/My Drive/Gresham Lab_Grace/france_satay/library_prep/france_saturation_scripts/grace_adapt_forBGI/on_hpc/cds_lengths/listAllGenes.txt", col_names = F, col_types = cols(X1 = col_character())) %>%
  dplyr::rename(id=X1)

#get file
get_gene_no_inserts=function(x, ver = "combined") {
  genes = read_csv(paste0(dir, '/hpc_output/', ver, '/', x), 
                   col_names = F, col_types = cols()) %>% 
    dplyr::rename(id=X1) %>% mutate(gene=NA, sample = str_sub(x,1,-26), insertions_per_gene=0, run = ver)
  genes$gene = yeast_r64_to_systematic(genes$id)
  genes = genes %>% mutate(gene = if_else(id == "ID= GRESHAMGFP", "CNV_reporter", gene))
  genes %>% dplyr::select(gene, sample, insertions_per_gene) %>% distinct()
}
no_insertion_files = list.files(path = paste0(dir, 'hpc_output/combined'), pattern = "genesWithNoInsertion.txt")
no_insertions = do.call(rbind, map(no_insertion_files,get_gene_no_inserts))
```

```{r combine insertion and no insertion data}
#combine
insertions_combined = ipg_df %>% 
  dplyr::select(-CDS, -run) %>%
  bind_rows(no_insertions) %>% distinct()
#reshape data
ipg = insertions_combined %>%
  spread(key=sample, value = insertions_per_gene) 

# normalize insertions per gene by dividing by library size
ipg_norm = ipg %>% 
  pivot_longer(cols = starts_with('1'), names_to = "sample", values_to = "insertions") %>%
  group_by(sample) %>%
  mutate(norm_inserts = insertions/sum(insertions))

```

### Use insertions per gene to investigate gene essentiality
We obtained a list of essential genes from the [Stanford yeast deletion collection](list of essential genes from yeast deletion collection). "Essential_ORFs.txt" downloaded Nov. 14, 2019.

We also generated a list of essential genes by using YeastMine and searching SGD for genes for which the null mutant has an ‘inviable’ phenotype, as in Michel et al.2017. List "Phenotype_Genes.csv" generated from YeastMine Nov. 14, 2019.

```{r essential gene lists}
#sgd
ess_sgd = read_csv('./files/Phenotype_Genes.csv', col_names = F, col_types = cols(
  X1 = col_character(),
  X2 = col_character(),
  X3 = col_character(),
  X4 = col_character(),
  X5 = col_character(),
  X6 = col_character(),
  X7 = col_character(),
  X8 = col_character(),
  X9 = col_logical(),
  X10 = col_character(),
  X11 = col_character(),
  X12 = col_character(),
  X13 = col_character(),
  X14 = col_character(),
  X15 = col_character(),
  X16 = col_character(),
  X17 = col_double(),
  X18 = col_character()
)) %>% 
  dplyr::rename(gene = X2) %>% dplyr::select(gene) %>%
  mutate(ess_sgd = "yes")
# deletion collection
ess_del = read_tsv('./files/Essential_ORFs.txt', col_names = T, comment = '=', col_types = cols(
  rec_num = col_double(),
  ORF_name = col_character(),
  deletion_alias = col_character(),
  gene_names = col_character(),
  UPTAG_sequence_20mer = col_character(),
  DNTAG_sequence_20mer = col_character()
)) %>%
  dplyr::select(ORF_name) %>% dplyr::rename(gene=ORF_name) %>%
  mutate(ess_del = "yes")
# combine
essential = ess_sgd %>% full_join(ess_del) %>% 
  mutate(ess_del = replace_na(ess_del, "no"), ess_sgd = replace_na(ess_sgd, "no")) %>%
  distinct() 
```

Here we are going to look at genes that have no insertions in any of the libraries (and so are probably essential in YPGal.)
```{r compare essential genes with insertion libraries}
ess_compare = ipg %>%
  full_join(essential %>% dplyr::select(gene, ess_del, ess_sgd), by = "gene")

essential_YPGal = ess_compare %>% 
  filter(`1657_1` == 0 & `1657_2` == 0 & `1728`== 0 & `1734` == 0& `1736` == 0 & `1740` == 0 & `1744` == 0 & `1747` == 0 & `1751` == 0)# & ess_del == "yes")
print(paste0("There are ", nrow(essential_YPGal), " genes that have no insertions in any of the libraries."))

#### write data for Table 1####
write_csv(essential_YPGal %>% mutate(common_name = yeast_systematic_to_common(gene)), "./figure_data/Table1.csv")
```

These are conditionally essential in YPGal for CNVs only - there are no insertions in any of the CNV strains, but 1657 has an insertion.
```{r essential YPGal CNV}
essentialCNV_YPGal = ess_compare %>% 
  filter(`1657_1` > 0 & `1657_2` > 0 & `1728` == 0  & `1734` == 0  & `1736` == 0  & `1740` == 0  & `1744` == 0 & `1747` == 0 & `1751` == 0)
print(paste0("There are ", nrow(essentialCNV_YPGal), " genes that may be conditionally essential in YPGal for CNV strains only."))
```
These genes are not essential in CNV strains (have 1 or more insertions in all of the CNV strains), but have no insertions in 1657 (suggesting they are essential in wild type)
```{r not essential YPGal CNV only}
NOTessentialCNV_YPGal = ess_compare %>% 
  dplyr::filter(`1657_1` == 0 & `1657_2` == 0  & `1728`> 0  & `1734` > 0  & `1736` > 0  & `1740` > 0  & `1744` > 0 & `1747` > 0 & `1751` > 0)
print(paste0("There are ", nrow(NOTessentialCNV_YPGal), " genes that are have no insertions in the euploid strain, but have insertions in the CNV strains."))
NOTessentialCNV_YPGal = NOTessentialCNV_YPGal %>%
  pivot_longer(cols = starts_with("1"), names_to = "strain", values_to = "insertions") %>%
  left_join(ipg_norm, by = c("gene" = "gene", "strain" = "sample", "insertions" = "insertions")) %>%
  mutate(common_name = yeast_systematic_to_common(gene))

#### write data for Figure 2A####
write_csv(NOTessentialCNV_YPGal, "./figure_data/Fig2A.csv")
```

### Do differential analysis on CNVs vs 1657

```{r}
#make counts matrix the way deseq wants it
#####this is for all cnv vs 1657####
counts = as.data.frame(ipg) %>% dplyr::select(-gene)
rownames(counts) = ipg$gene
#make col data frame describing exp
coldata = data.frame(type = c("wt", "wt", "cnv", "cnv", "cnv", "cnv", "cnv","cnv", "cnv"), sample = c('1657_1','1657_2','1728','1734', '1736', '1740', '1744','1747','1751'), row.names = colnames(counts))
coldata$type = factor(coldata$type, levels = c('wt', 'cnv'))
#make deseq data
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = coldata,
                              design = ~ type)
dds <- DESeq(dds)
res <- results(dds, alpha=0.05)
summary(res)

#### write data for Figure 2B####
t=as_tibble(res) %>%
            mutate(gene = rownames(res), sig = padj < 0.05) %>% 
            filter(!is.na(padj)) %>% #remove rows with baseMean = 0 (meaning zero insertions) for plotting
  mutate(common_name = yeast_systematic_to_common(gene))
write_csv(t, "./figure_data/Fig2B.csv")


# do GSEA based on log2FC
## assume 1st column is ID
## 2nd column is FC

## feature 1: numeric vector
geneList_allCNVdiff = res$log2FoldChange[-1] #remove GRESHAMGFP
## feature 2: named vector
gene.df <- bitr(rownames(res)[-1], fromType = "ORF",
        toType = "ENTREZID",
        OrgDb = org.Sc.sgd.db::org.Sc.sgd.db)
names(geneList_allCNVdiff) = gene.df$ENTREZID
## feature 3: decreasing order
geneList_allCNVdiff = sort(geneList_allCNVdiff, decreasing = TRUE)
ego <- gseGO(geneList     = geneList_allCNVdiff,
             OrgDb        = org.Sc.sgd.db::org.Sc.sgd.db,
             keyType = "ENTREZID",
              ont          = "BP",
              nPerm        = 1e6,
              minGSSize    = 10,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              pAdjustMethod="fdr",
              verbose      = FALSE,
             seed = 1)


gsea_out = as_tibble(ego@result) %>% mutate(sample = "All_CNV") %>% select(sample, ID,Description, setSize, enrichmentScore, NES, p.adjust,rank, pvalue, core_enrichment) %>% arrange(sample,-enrichmentScore)
```
### Do differential analysis and GSEA on each individual CNV strain vs 1657
```{r}
significant_by_strain = NULL

for(strain in c("1728", "1734", "1736", "1740", "1744", "1747", "1751")) {
  counts = as.data.frame(ipg) %>% dplyr::select(`1657_1`, `1657_2`, strain)
  rownames(counts) = ipg$gene
  coldata = data.frame(type = c("wt", "wt", "cnv"), sample = c('1657_1','1657_2',strain), row.names = colnames(counts))
  coldata$type = factor(coldata$type, levels = c('wt', 'cnv'))
  dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = coldata,
                              design = ~ type)
  dds <- DESeq(dds)
  res <- results(dds, alpha=0.05)
  significant_by_strain = bind_rows(significant_by_strain, 
                                    as_tibble(res) %>% 
                                      mutate(gene = rownames(res), strain = strain) %>% 
                                      filter(padj < 0.05))
  geneList = res$log2FoldChange[-1] #remove GRESHAMGFP
  gene.df <- bitr(rownames(res)[-1], fromType = "ORF",
        toType = "ENTREZID",
        OrgDb = org.Sc.sgd.db::org.Sc.sgd.db)
  names(geneList) = gene.df$ENTREZID
  geneList = sort(geneList, decreasing = TRUE)
  ego <- gseGO(geneList     = geneList,
             OrgDb        = org.Sc.sgd.db::org.Sc.sgd.db,
             keyType = "ENTREZID",
              ont          = "BP",
              nPerm        = 1e6,
              minGSSize    = 10,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              pAdjustMethod="fdr",
              verbose      = FALSE,
             seed = 1)
  gsea_out = bind_rows(gsea_out, as_tibble(ego@result) %>% mutate(sample = strain))
}
#### write data for Figure 2C - but it needs to be summarized by REVIGO first####
gsea_out = gsea_out %>% select(sample,ID,Description, enrichmentScore, setSize, NES, p.adjust,rank, pvalue, core_enrichment) %>% arrange(sample,-enrichmentScore)
write_tsv(gsea_out, "./figure_data/Fig2C_to_summarize.tsv")

# gather together REVIGO summarized data
revigo = NULL
for(set in c("1728", "1734", "1736", "1740", "1744", "1747", "1751", "AllCNV")){
  revigo = bind_rows(revigo, read_csv(paste0("./figure_data/REVIGO", set, ".csv"), 
                                      comment = "#") %>%
    select(!starts_with("plot"), -description) %>% mutate(userVal_2 = if_else(is.na(userVal_2), "All_CNV", as.character(userVal_2))))
}

gsea_out_final = gsea_out %>% full_join(revigo, by = c("ID" = "term_ID", "sample" = "userVal_2")) %>%
  mutate(eliminated = if_else(is.na(eliminated), 0, eliminated)) %>% # zero is keep, 1 is eliminate
  mutate(group = case_when(Description == "mitochondrial translation" ~ "translation",
                           Description == "mitochondrial gene expression" ~ "mitochondrial function",
                           Description == "cytoplasmic translation" ~ "translation",
                           Description == "cytochrome complex assembly" ~ "mitochondrial function",
                           Description == "translational elongation" ~ "translation",
                           Description == "ribosomal large subunit biogenesis" ~ "ribsome biogenesis",
                           Description == "nucleic acid phosphodiester bond hydrolysis" ~ "",
                           Description == "organelle localization" ~ "cellular component organization",
                           Description == "protein localization to chromosome, centromeric region" ~ "",
                           Description == "aerobic electron transport chain" ~ "mitochondrial function",
                           Description == "ribonucleoprotein complex subunit organization" ~ "complexes",
                           Description == "ncRNA processing" ~ "RNA processing",
                           Description == "ribosome biogenesis" ~ "ribosome biogenesis",
                           Description == "ncRNA metabolic process" ~ "RNA processing",
                           Description == "transposition" ~ "",
                           Description == "transposition, RNA-mediated" ~ "",
                           Description == "Golgi vesicle transport" ~ "",
                           Description == "organelle membrane fusion" ~ "cellular component organization",
                           Description == "rRNA 5'-end processing" ~ "ribosome biogenesis",
                           Description == "ncRNA export from nucleus" ~ "RNA processing",
                           Description == "mRNA metabolic process" ~ "RNA processing",
                           Description == "electron transport chain" ~ "mitochondrial function",
                           Description == "cellular oxidant detoxification" ~ "",
                           Description == "intracellular protein transport" ~ "",
                           Description == "regulation of cellular component organization" ~ "cellular component organization",
                           Description == "DNA recombination" ~ "DNA replication & repair",
                           Description == "lipid biosynthetic process" ~ "",
                           Description == "cell division" ~ "cell cycle",
                           Description == "mitotic cell cycle" ~ "cell cycle",
                           Description == "posttranscriptional regulation of gene expression" ~ "",
                           Description == "ribonucleoprotein complex assembly" ~ "complexes",
                           Description == "chromosome segregation" ~ "DNA replication & repair",
                           Description == "RNA phosphodiester bond hydrolysis" ~ "RNA processing",
                           Description == "microtubule-based process" ~ "",
                           Description == "peptidyl-lysine modification" ~ "post-translational regulation",
                           Description == "DNA-templated transcription, elongation" ~ "transcription",
                           Description == "mitotic nuclear division" ~ "cell cycle",
                           Description == "organic hydroxy compound biosynthetic process" ~ "",
                           Description == "protein maturation" ~ "post-translational regulation",
                           Description == "DNA-templated transcription, initiation" ~ "transcription",
                           Description == "transcription initiation from RNA polymerase II promoter" ~ "transcription",
                           Description == "regulation of DNA-templated transcription, elongation" ~ "transcription",
                           Description == "DNA-templated transcription, termination" ~ "transcription",
                           Description == "amino acid activation" ~ "",
                           Description == "membrane docking" ~ "cellular component organization",
                           Description == "organelle localization by membrane tethering" ~ "cellular component organization",
                           Description == "positive regulation of proteolysis" ~ "post-translational regulation",
                           Description == "polyadenylation-dependent snoRNA 3'-end processing" ~ "RNA processing",
                           Description == "proteasomal ubiquitin-independent protein catabolic process" ~ "post-translational regulation",
                           Description == "snRNA processing" ~ "RNA processing",
                           Description == "exocytic process" ~ "",
                           Description == "vesicle docking involved in exocytosis" ~ "",
                           Description == "mRNA processing" ~ "RNA processing"))

# write final data for Figure 2C and Supplementary Table X
write_csv(gsea_out_final, "./figure_data/Fig2C.csv")

#### write data for Figure 2D####
significant_by_strain$common_name = yeast_systematic_to_common(significant_by_strain$gene)

### I NEED TO ADD IN ANNOTATION FOR COPY NUMBER ####

write_csv(significant_by_strain, "./figure_data/Fig2D.csv")
```

### Do differential analysis on CNVs vs 1657 - windows based approach
This is to account for non-coding regions as well as protein domains
```{r}
#make counts matrix the way deseq wants it
window_counts = cov_df %>% 
  pivot_wider(names_from = "sample", values_from = "coverage") %>%
  mutate(location = paste(chromosome, bin_start, sep = "_")) %>%
  select(-chromosome, -bin_start, -bin_end)

#####this is for all cnv vs 1657####
counts = as.data.frame(window_counts) %>% dplyr::select(-location)
rownames(counts) = window_counts$location
#make col data frame describing exp
coldata = data.frame(type = c("wt", "wt", "cnv", "cnv", "cnv", "cnv", "cnv","cnv", "cnv"), sample = c('1657_1','1657_2','1728','1734', '1736', '1740', '1744','1747','1751'), row.names = colnames(counts))
coldata$type = factor(coldata$type, levels = c('wt', 'cnv'))
#make deseq data
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = coldata,
                              design = ~ type)
dds <- DESeq(dds)
res <- results(dds, alpha=0.05)
windows_significant_by_strain = as_tibble(res) %>% 
                                      mutate(location = rownames(res), strain = "All_CNV") %>% 
                                      filter(padj < 0.05)

for(strain in c("1728", "1734", "1736", "1740", "1744", "1747", "1751")) {
  counts = as.data.frame(window_counts) %>% dplyr::select(`1657_1`, `1657_2`, strain)
  rownames(counts) = window_counts$location
  coldata = data.frame(type = c("wt", "wt", "cnv"), sample = c('1657_1','1657_2',strain), row.names = colnames(counts))
  coldata$type = factor(coldata$type, levels = c('wt', 'cnv'))
  dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = coldata,
                              design = ~ type)
  dds <- DESeq(dds)
  res <- results(dds, alpha=0.05)
  windows_significant_by_strain = bind_rows(windows_significant_by_strain, 
                                    as_tibble(res) %>% 
                                      mutate(location = rownames(res), strain = strain) %>%
                                      filter(padj < 0.05))
}
```


## Relative frequency of insertions in CNV regions vs 1657

The CNV strains were characterized in Lauer et al. 2018. S4 Table. I will subset out the genes inside the CNV region for each strain.

```{r cnv regions}
strain_info = read_csv('journal.pbio.3000069.s021.csv', col_types = cols(
  `Clone ID` = col_character(),
  `Population ID` = col_character(),
  `Lab collection ID` = col_character(),
  `CNV Reporter` = col_character(),
  Selection = col_character(),
  `Generation Isolated` = col_double(),
  `Median fluorescence` = col_double(),
  `RD estimation of GAP1 copy number` = col_double(),
  `RD estimation of DUR3 copy number` = col_double(),
  `RD estimation of  CUP1 copy number` = col_double(),
  `RD estimation of  rDNA copy number` = col_double(),
  `CNV Size (kb)` = col_character(),
  `Left Boundary` = col_double(),
  `Left Gene` = col_character(),
  `Left Gene Class` = col_character(),
  `Right Boundary` = col_double(),
  `Right Gene` = col_character(),
  `Right Gene Class` = col_character()
)) %>% filter(`Lab collection ID` %in% c('DGY1728','DGY1734', 'DGY1736', 'DGY1740','DGY1744', 'DGY1747', 'DGY1751'))
#get gff
gff = read_tsv(paste0(dir,'/onHPC/cds_lengths/GCF_000146045.2_R64_genomic_GAP1.gff'),
                comment='#', col_names = F, col_types = cols(
  X1 = col_character(),
  X2 = col_character(),
  X3 = col_character(),
  X4 = col_double(),
  X5 = col_double(),
  X6 = col_character(),
  X7 = col_character(),
  X8 = col_character(),
  X9 = col_character()))
#get genes with CNV region for each strain
genes_region_1728 = gff %>% filter(X3=="CDS",X1 == 'NC_001143.9') %>% 
  separate(X9, into=c('id'), sep=';') %>%
  mutate(gene = NA) %>% dplyr::select(id, gene)

genes_region_1734 = gff %>% filter(X3=="CDS", X1 == 'NC_001143.9', 
                                   X4 >= strain_info$`Left Boundary`[strain_info$`Lab collection ID`=='DGY1734'], X5 <= strain_info$`Right Boundary`[strain_info$`Lab collection ID`=='DGY1734']) %>% 
  separate(X9, into=c('id'), sep=';') %>%
  mutate(gene = NA) %>% dplyr::select(id, gene)

genes_region_1736 = gff %>% filter(X3=="CDS", X1 == 'NC_001143.9', 
                                   X4 >= strain_info$`Left Boundary`[strain_info$`Lab collection ID`=='DGY1736'], X5 <= strain_info$`Right Boundary`[strain_info$`Lab collection ID`=='DGY1736']) %>% 
  separate(X9, into=c('id'), sep=';') %>%
  mutate(gene = NA) %>% dplyr::select(id, gene)

genes_region_1740 = gff %>% filter(X3=="CDS", X1 == 'NC_001143.9', 
                                   X4 >= strain_info$`Left Boundary`[strain_info$`Lab collection ID`=='DGY1740'], X5 <= strain_info$`Right Boundary`[strain_info$`Lab collection ID`=='DGY1740']) %>% 
  separate(X9, into=c('id'), sep=';') %>%
  mutate(gene = NA) %>% dplyr::select(id, gene)

genes_region_1744 = gff %>% filter(X3=="CDS", X1 == 'NC_001143.9', 
                                   X4 >= strain_info$`Left Boundary`[strain_info$`Lab collection ID`=='DGY1744'], X5 <= strain_info$`Right Boundary`[strain_info$`Lab collection ID`=='DGY1744']) %>% 
  separate(X9, into=c('id'), sep=';') %>%
  mutate(gene = NA) %>% dplyr::select(id, gene)

genes_region_1747 = gff %>% filter(X3=="CDS", X1 == 'NC_001143.9', 
                                   X4 >= strain_info$`Left Boundary`[strain_info$`Lab collection ID`=='DGY1747'], X5 <= strain_info$`Right Boundary`[strain_info$`Lab collection ID`=='DGY1747']) %>% 
  separate(X9, into=c('id'), sep=';') %>%
  mutate(gene = NA) %>% dplyr::select(id, gene)

genes_region_1751 = gff %>% filter(X3=="CDS", X1 == 'NC_001143.9', 
                                   X4 >= strain_info$`Left Boundary`[strain_info$`Lab collection ID`=='DGY1751'], X5 <= strain_info$`Right Boundary`[strain_info$`Lab collection ID`=='DGY1751']) %>% 
  separate(X9, into=c('id'), sep=';') %>%
  mutate(gene = NA) %>% dplyr::select(id, gene)

####DO THIS BETTER####
for(i in 1:nrow(genes_region_1728)) {
  genes_region_1728$gene[i] = change_id_tosystematic(genes_region_1728$id[i], id_db)
}
for(i in 1:nrow(genes_region_1734)) {
  genes_region_1734$gene[i] = change_id_tosystematic(genes_region_1734$id[i], id_db)
}
for(i in 1:nrow(genes_region_1736)) {
  genes_region_1736$gene[i] = change_id_tosystematic(genes_region_1736$id[i], id_db)
}
for(i in 1:nrow(genes_region_1740)) {
  genes_region_1740$gene[i] = change_id_tosystematic(genes_region_1740$id[i], id_db)
}
for(i in 1:nrow(genes_region_1744)) {
  genes_region_1744$gene[i] = change_id_tosystematic(genes_region_1744$id[i], id_db)
}
for(i in 1:nrow(genes_region_1747)) {
  genes_region_1747$gene[i] = change_id_tosystematic(genes_region_1747$id[i], id_db)
}
for(i in 1:nrow(genes_region_1751)) {
  genes_region_1751$gene[i] = change_id_tosystematic(genes_region_1751$id[i], id_db)
}
```

### Plot relative frequency of insertions in CNV regions 
```{r boxplots rel freq inserts CNV region}

freq_region_kb_gene = ipg_norm %>%
  mutate(keep = case_when(sample %in% c("1657_1", "1657_2", "1728") & gene %in% genes_region_1728$gene ~ "yes",
                          sample %in% c("1734") & gene %in% genes_region_1734$gene ~ "yes",
                          sample %in% c("1736") & gene %in% genes_region_1736$gene ~ "yes",
                          sample %in% c("1740") & gene %in% genes_region_1740$gene ~ "yes",
                          sample %in% c("1744") & gene %in% genes_region_1744$gene ~ "yes",
                          sample %in% c("1747") & gene %in% genes_region_1747$gene ~ "yes",
                          sample %in% c("1751") & gene %in% genes_region_1751$gene ~ "yes")) %>%
  filter(keep == "yes") %>%
  select(-keep) %>%
  mutate(essential = if_else(gene %in% ess_del$gene, "Essential", "Non-essential"))

ggplot(freq_region_kb_gene, aes(sample, norm_inserts, color = essential)) +
  geom_boxplot(outlier.shape = NA) + 
  geom_beeswarm(aes(alpha=0.3), cex = 0.3, dodge.width = 0.75) +
  scale_color_manual(values = c("#8dd3c7","#bebada")) +
  ylab("Normalized insertions per gene w/in CNV region") 

```

